### ISSUE TITLE
Very poor perf on a chain example

### ISSUE THREAD

**Issue by nrc** [rust-lang-nursery/rustfmt#1849]
I'm not sure exactly what is going on here, but there are a lot of recursive calls on chains and the stack is 180 deep pretty much all in rustfmt code. The test case is [rls-analysis](https://github.com/nrc/rls-analysis) which has some pretty deep closure/chain combos. I haven't made a reduced test case yet.

**Comment by topecongiro** [comment]
E.g.
```rust
    self.with_analysis(|a| {
        a.for_each_crate(|c| {
            c.def_id_for_span
                .get(span)
                .and_then(|id| if c.defs.contains_key(id) {
                    Some(id)
                } else {
                    None
                })
                .cloned()
        })
    })
```
when formatting the above snippet, `rewrite_method_call()` is called over 5000 times. I am not sure why.

**Comment by topecongiro** [comment]
I think the root problem here is in closure + function call rather than chain.

**Comment by topecongiro** [comment]
rustfmt uses 'try in several ways and pick the best one' strategy in various places.
(E.g. the last argument for function call, rhs of assignment, the last element of chain and others)
This is harmless when the expression is simple, but when they get combined like in the above example, it gets exponentailly worse.

**Comment by bluss** [comment]
A couple of the crates I tried now take longer to reformat than to build. itertools 0.6.1 (ea7697a) was one, using 10 seconds.

**Comment by topecongiro** [comment]
Closed via #1861.
